### PR TITLE
Added inspector reports to the end game score screen

### PR DESCRIPTION
**Pull request by TobleroneSwordfish**
[FEATURE]

## About the PR 
The Inspector now starts with a special clipboard, anything written on it will be displayed as a report at the top of the ticket window on round end. Can handle multiple Inspectors, inspector clipboards with no owner set, and culling blank pages automatically.

report

The section headers are the names of the pieces of paper.

EDIT: the report is now shown above tickets, as it will not be there every round so players probably wouldn't look for it under a list of 7 tickets for underage drinking.

The Inspector's special clipboard in game :
image
image
The inhand icon is just a tint of the original clipboard sprite so looks a little off, but I don't really want to add 16 new sprites to fix that.
## Why's this needed? 
Inspectors are meant to write reports, sometimes they do, but mostly no-one ever gets to see them besides the captain who doesn't care and some guy in the bar who isn't listening.

Pros:
- Inspectors now have a reason to write reports
- Having a little description of some of the things you didn't see happen in a round would be cute
- Highlights the HoP's continued incompetence

Cons:
- None
- People will put the navy seal copypasta in it

## Changelog

```changelog
(u)LeahTheTech
(+)The Inspector's report will now be displayed at the end of the round. The contents of that ominous clipboard are now much more important.
```

### CODE REVIEW

File: code/procs/scorestats.dm [hunk 1]
```diff
@@ -352,6 +355,33 @@ var/datum/score_tracker/score_tracker
 
 		return jointext(., "")
 
+	//I feel like I may be reinventing the wheel here, but here's a recursive wheel
+	//recursively finds all items of type in container
+	proc/find_items_in(obj/container, type)
```
Comment (Zonespace27) [inline code/procs/scorestats.dm:360]: Fairly sure there's a proc that does this already.

File: code/procs/scorestats.dm [hunk 2]
```diff
@@ -352,6 +355,33 @@ var/datum/score_tracker/score_tracker
...
+
+	proc/get_inspector_report()
+		var/report = ""
+		//this might be a really stupid way of finding all players of a specific job, please tell me if it is
+		for (var/mob/living/carbon/human/player in mobs)
+			if (player.job != "Inspector")
+				continue
+			report += "<B>Inspector [player.real_name]'s report</B><BR><HR>"
+			//find all clipboards they have on them (probably just one, but may as well check)
+			var/list/clipboards = find_items_in(player, /obj/item/clipboard)
```
Comment (Zonespace27) [inline code/procs/scorestats.dm:377]: The player is not an obj, but "find_items_in" typecasts that argument to an obj.

File: code/procs/scorestats.dm [hunk 3]
```diff
@@ -400,7 +430,7 @@ var/datum/score_tracker/score_tracker
 	src.Browse(score_tracker.score_text, "window=roundscore;size=500x700;title=Round Statistics")
 
 /mob/proc/showtickets()
-	if(!data_core.tickets.len && !length(data_core.fines)) return
+	if(!data_core.tickets.len && !length(data_core.fines) && length(score_tracker.inspector_report) == 0) return
```
Comment (Zonespace27) [inline code/procs/scorestats.dm:433]:
```suggestion
	if(!data_core.tickets.len && !length(data_core.fines) && !length(score_tracker.inspector_report)) return
```

File: code/procs/scorestats.dm [hunk 4]
```diff
@@ -352,6 +355,33 @@ var/datum/score_tracker/score_tracker
...
+	proc/get_inspector_report()
+		var/report = ""
+		//this might be a really stupid way of finding all players of a specific job, please tell me if it is
+		for (var/mob/living/carbon/human/player in mobs)
+			if (player.job != "Inspector")
+				continue
+			report += "<B>Inspector [player.real_name]'s report</B><BR><HR>"
+			//find all clipboards they have on them (probably just one, but may as well check)
+			var/list/clipboards = find_items_in(player, /obj/item/clipboard)
+			for (var/obj/item/clipboard/clipboard as anything in clipboards)
```
Comment (Zonespace27) [inline code/procs/scorestats.dm:378]: If player has multiple clipboards, multiple reports will be shown. I would suggest a "break" in the for loop after it's found and done 1 clipboard.

File: code/procs/scorestats.dm (hunk 1 above)
Comment (TobleroneSwordfish) [reply code/procs/scorestats.dm:360]: I thought so but couldn't find it, any idea what it's called? Searching global procs for "find" etc hasn't turned it up.

File: code/procs/scorestats.dm (hunk 4 above)
Comment (TobleroneSwordfish) [reply code/procs/scorestats.dm:378]: I had it written like that originally, but thought that a player randomly picking up an empty clipboard isn't too uncommon an event and then it's a 50/50 chance whether their report actually gets shown. It shouldn't show up as a separate report exactly, just add on to the end of the current one.

File: code/procs/scorestats.dm (hunk 2 above)
Comment (TobleroneSwordfish) [reply code/procs/scorestats.dm:377]: Good point, is there a way I should type this to ensure that .contents exists? I'm not sure which base class it comes from.

File: code/procs/scorestats.dm (hunk 1 above)
Comment (Zonespace27) [reply code/procs/scorestats.dm]: `/mob/proc/get_all_items_on_mob()`

File: code/procs/scorestats.dm (hunk 2 above)
Comment (Zonespace27) [reply code/procs/scorestats.dm:377]: typecast it to `atom` and restrict it to obj/mob

File: code/procs/scorestats.dm (hunk 4 above)
Comment (TobleroneSwordfish) [reply code/procs/scorestats.dm:378]: Also empty papers/clipboards won't show up on the endscreen

File: code/procs/scorestats.dm [hunk 5]
```diff
@@ -400,7 +419,7 @@ var/datum/score_tracker/score_tracker
 	src.Browse(score_tracker.score_text, "window=roundscore;size=500x700;title=Round Statistics")
 
 /mob/proc/showtickets()
-	if(!data_core.tickets.len && !length(data_core.fines)) return
+	if(!data_core.tickets.len && !length(data_core.fines) && !length(score_tracker.inspector_report)) return
```
Comment (Zonespace27) [inline code/procs/scorestats.dm:422]:
```suggestion
	if(!length(data_core.tickets) && !length(data_core.fines) && !length(score_tracker.inspector_report)) return
```
May as well

File: code/procs/scorestats.dm [hunk 6]
```diff
@@ -352,6 +355,22 @@ var/datum/score_tracker/score_tracker
...
+			report += "<B>Inspector [player.real_name]'s report</B><BR><HR>"
+			//find all clipboards they have on them (probably just one, but may as well check)
+			var/list/items = player.get_all_items_on_mob()
+			for (var/obj/item/clipboard/clipboard in items)
+				for(var/obj/item/paper/paper in clipboard.contents)
+					if (paper.name != "paper")
+						report += "<B>[paper.name]</B>"
+					report += paper.info
+		return report
+
```
Comment (frawhst) [inline code/procs/scorestats.dm:373]: Don't like this- inefficient to loop through all mobs (and then all items on the inspector), and creates the issue where dropping or losing the clipboard close to the end of the round means your report doesn't show up. Would rather you track the clipboards themselves- maybe create a `clipboard/inspector` subtype, track that (with `START_TRACKING`), then iterate through all inspector clipboards at the end of the round and add their inspection notes to the ticket report. Would ensure the notes show up even if the inspector leaves their clipboard somewhere, and much more efficient.

That implementation does mean the inspector can't replace their one Important clipboard if they lose it, but I don't really mind that- could just name it "Inspector's Clipboard" so they can distinguish it from others. Also allows people to gank the inspector's clipboard and pin silly things on it, which will then show up at the end of the round regardless of who's holding it.

**Review by frawhst — CHANGES_REQUESTED** [review]
Generally, cool feature. Would like to have this.

File: code/procs/scorestats.dm (hunk 6 above)
Comment (TobleroneSwordfish) [reply code/procs/scorestats.dm]: Have now implemented pretty much exactly this, thanks for the pointers!

**Comment by TobleroneSwordfish** [comment]
clipboard_inspector clipboard_inspector_with_overlays

an attempt at making a unique sprite for the clipboard, though I am not an artist and the base clipboard sprite could probably do with an update anyway

**Review by frawhst — APPROVED** [review]

**Pull request merged**